RFC.md

```diff
@@ -0,0 +1,413 @@
+        return out
+```
+
+### TransformerCrossAttentionLayer
+New layer. Discussion in the comments.
+```python
+class TransformerCrossAttentionLayer(nn.Module):
+    def __init__(
+		...
+    ) -> None:
+        super().__init__()
+        assert (
+            attn.pos_embeddings is None
+        ), "Positions are not computed for encoder inputs"
+        ...
+
+    def _skip_mask(self, mask: Optional[Tensor]) -> Optional[Tensor]:
+	    """Some tokens in x may not attend to any encoder inputs
+        due to the cross attention mask (encoder_mask). This results in
+        a full row of the attention matrix being masked out.
+
+        In the example below, the word "the" is masked from every embedding.
+
+        .. code-block:: text
+
+            |emb||emb||emb|
+        |The| x    x    x
+        |red|      x
+        |car| x
+
+        This results in no inputs into the softmax layer which causes a NaN.
+        The skip mask removes the output of the attention module and
+        mlp resulting in the token being skipped.
+
+        """
+		# There is more discussion around this here
+		# https://gist.github.com/drisspg/547648ded500d078961b7a3b3f11c310
+		# We might be able to optimize this approach more
+        if mask is None:
+            return None
+        if mask.dtype == torch.bool:
+            mask = ~mask
+        else:
+            mask = torch.isneginf(mask)
+        mask = torch.all(mask, dim=-1, keepdim=True)
+        return mask
+
+    def forward(
+        self,
+        x: Tensor,
+        *,
+        encoder_input: Optional[Tensor] = None,
+        encoder_mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+        **kwargs,
+    ) -> Tensor:
+        # Cross attention only layer (see red only) has no
+        # purpose if no conditional input is provided
+        if encoder_input is None:
+            return x
+
+        # A mask of tokens (x) with no encoder_input
+        skip_mask = self._skip_mask(encoder_mask)
+
+        # Input tensor and attention output have the same shape
+        # [b, s, d]
+        # Norm applied before self-attention
+        attn_out = self.attn(
+            self.attn_norm(x),
+            encoder_input,
+            mask=encoder_mask,
+            input_pos=input_pos
+        )
+        if skip_mask is not None:
+            attn_out.masked_fill_(skip_mask, 0)
+
+        # Residual connection; shape: [batch_size, seq_length, embed_dim]
+        h = self.attn_scale(attn_out) + x
+
+        # Norm applied before the feedforward layer
+        mlp_out = self.mlp(self.mlp_norm(h))
+        if skip_mask is not None:
+            mlp_out.masked_fill_(skip_mask, 0)
+
+        # Residual connection; shape: [batch_size, seq_length, embed_dim]
+        out = h + self.mlp_scale(mlp_out)
+        return out
+```
+@Chillee the `_skip_mask` is an example of a fully masked out row. Is applying the mask this way the most efficient way to handle this?
```

Comment:
To confirm, can we fold `_skip_mask` into flex attention after https://github.com/pytorch/pytorch/pull/131012?

Reply:
We can't get rid of it. flex attention can remove the need for applying it to the attention output but we still need it then for the feedforward.